### DISCUSSION TITLE
MERN App, Frontend and Backend deployed separately, Cannot get Frontend and Backend communication

### DISCUSSION THREAD

**Discussion by kristo-baricevic** [vercel/community#3810] [Help]
### Summary

I have a basic CRUD app built as a MERN stack with the frontend deployed to Vercel and the backend deployed to Fly.io.

I have not been able to figure out to set up the routing to get the axios methods to go through to the endpoints. 

My current setup was an action in a redux-toolkit slice being called, which invokes an axios function to post user data through a backend.js and send a temporary sign in to the backend. 

I have debugging logs in the backend.js methods that aren't running. As for as I can tell it's never getting to the backend. I've tried playing around with routes in a vercel.json file, and I've moved the api/backend.js file around to try different places in the directory. I came to this method from a suggestion in an old forum post elsewhere. 

I have tried to understand the vercel docs but i was not sure where to start because it seemed to go in different directions based on variables. 

would love any help! this app worked locally and i have been spending a week trying to deploy :/

### Example

https://vercel.com/kristo-baricevic/kristos-closet

### Steps to Reproduce

This is the function where the code dies:

export const loginAnonymous = createAsyncThunk(
  "user/loginAnonymous",
  async () => {
    try {
      console.log("inside loginAnonymous");
      const response = await axios.post('/api/loginAnonymous'); //this should send it to backend.js
      const isAuthenticated = response.data.isAuthenticated;
     //code to set token
      return { isAuthenticated, user };
    } catch (error) {
      throw error;
    }
  }
);

backend.js route:

router.post('/loginAnonymous', async (req, res) => {
  console.log("inside backend.js try"); // this never runs
  try {
    console.log("inside backend.js try");
    const response = await axios.post(`https://closet-app-backend.fly.dev/api/loginAnonymous`);
    res.status(response.status).json(response.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Internal server error'});
  }
});

**Comment by amyegan** [comment]
Hey @kristo-baricevic. You may need to [enable CORS](https://vercel.com/guides/how-to-enable-cors) or [reverse proxy with rewrites](https://vercel.com/guides/vercel-reverse-proxy-rewrites-external)

**Reply by kristo-baricevic** [reply]
Hi Amy - Thanks for the response. 

In response to CORS, I currently have CORS enabled in a server.js file that runs my backend deployment. Do I need it enabled somewhere on the frontend as well? I do not currently have a next.config file as one of the options suggests in that guide. Do I need that file? Would it go in the root directory? The guide suggests adding an allowCors method but I can't tell in what file that method would go.

In regards to the reverse proxies and rewrite, that was one thing I was considering but it's not clear what the methodology is. In the following sample code from a vercel.json file:

"source": "/blog",
      "destination": "https://acme.com/blog"
      
 is this taking a /blog endpoint and redirecting it to a separate site, https://acme.com/blog?
 
 And if this is the case, is including that in a vercel.json all that would be necessary to send a handshake to the backend?

**Reply by kristo-baricevic** [reply]
i have been working your suggestion, here is the current vercel.json:

`{
    "headers": [
        {
          "source": "/api/(.*)",
          "headers": [
            { "key": "Access-Control-Allow-Credentials", "value": "true" },
            { "key": "Access-Control-Allow-Origin", "value": "*" },
            { "key": "Access-Control-Allow-Methods", "value": "GET,OPTIONS,PATCH,DELETE,POST,PUT" },
            { "key": "Access-Control-Allow-Headers", "value": "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version" }
          ]
        }
      ],
    "rewrites": [
        { 
            "source": "/api/images/(.*)", 
            "destination": "https://closet-app-backend.fly.dev/api/images" ,
            "methods": ["GET", "PUT", "DELETE"]
        },
        {
             "source": "/api/upload/(.*)", 
             "destination": "https://closet-app-backend.fly.dev/api/upload",
             "methods": ["POST"] 
        },
        { 
            "source": "/api/register/(.*)", 
            "destination": "https://closet-app-backend.fly.dev/api/register",
            "methods": ["POST"]
        },
        { 
            "source": "/api/login/(.*)",
            "destination": "https://closet-app-backend.fly.dev/api/login",
            "methods": ["POST"]
         },
        { 
            "source": "/api/logout/(.*)", 
            "destination": "https://closet-app-backend.fly.dev/api/logout",
            "methods": ["POST"]
        },
        { 
            "source": "/api/loginAnonymous/(.*)", 
            "destination": "https://closet-app-backend.fly.dev/api/loginAnonymous", 
            "methods": ["POST"] 
        },
        { 
            "source": "/api/current/(.*)", 
            "destination": "https://closet-app-backend.fly.dev/api/current",
            "methods": ["GET"]
        }
    ]
}`
  
and i have been sending axios methods like so: 

`// Create an async thunk for loginAnonymous
export const loginAnonymous = createAsyncThunk(
  "user/loginAnonymous",
  async () => {
    try {
      console.log("inside loginAnonymous");
      const response = await axios.post('/api/loginAnonymous');
      const isAuthenticated = response.data.isAuthenticated;
      console.log("after axios.post with value", isAuthenticated);
      const user = response.data.user;
      const token = response.data.token;
      localStorage.setItem('token', token);
      return { isAuthenticated, user };
    } catch (error) {
      throw error;
    }
  }
);`

**Reply by MattHammond94** [reply]
Hi Kristo, 

Did you manage to resolve the issue you were facing? I'm encountering a similar issue whereby I have deployed my backend express server (And can send REQ's to it using postman to get a correct RES.)

I have deployed my frontend vite(react) application separately. In my frontend application I have been using REDUX toolkit to send REQ's to the backend and in development configured vite to use a proxy however now deployed my REQ's are not being sent to my backend and I cannot seem to locate the issue as to why.

I have enabled CORS on the backend server simply to check if this was causing the issue however it is not and my logs seem to indicate that the REQ's are being made to the frontend URL as opposed to the backend. 

Any help or a nudge in the right direction would be greatly appreciated!

Thanks,
Matt
